File: framework/lib/property_changed_event.dart
part of uxml;

/**
 * Defines event class for changes to property of an object.
 *
 * @author sanjayc@ (Sanjay Chouksey)
 */
class PropertyChangedEvent extends EventArgs {

  /** The property change event definition. */
  static EventDef changeEvent;
  static void initPropertyChangedEvent() {
    changeEvent = new EventDef("change", Route.DIRECT);
  }

  /**
   * Queue used for object pooling. User calls
   * static create and release to access shared instances.
   */
  static List<PropertyChangedEvent> _objectPool;

  /** The property that changed. */
  Object property;

  /** The old value of the property. */
  Object oldValue;

  /** The new value of the property. */
  Object newValue;

  /** Constructs PropertyChangedEvent. */
  PropertyChangedEvent(Object source,
                       Object propertyKey,
                       Object oldValue,
                       Object newValue) : super(source) {
    this.property = propertyKey;
    this.oldValue = oldValue;
    this.newValue = newValue;
  }

  /**
   * Returns a PropertyChangedEvent by reusing an instance from an objectPool
   * or by creating a new instance.
   * @param source Source element of the property change event.
   * @param property Contains property definition or key for the change event.
   * @param oldValue Contains old property value.
   * @param newValue Container new property value.
   */
  static PropertyChangedEvent create(Object source, Object propertyKey,
      Object oldValue, Object newValue) {
    if (_objectPool == null) {
      _objectPool = <PropertyChangedEvent>[];
    }
    if (_objectPool.length == 0) {
      return new PropertyChangedEvent(source, propertyKey,
          oldValue, newValue);
    }
    PropertyChangedEvent propEvent = _objectPool.removeLast();
    propEvent.source = source;
    propEvent.currentSource = source;
    propEvent.property = propertyKey;
    propEvent.oldValue = oldValue;
    propEvent.newValue = newValue;
    return propEvent;
  }

  /**
   * Returns a PropertyChangedEvent instance to the shared pool to reduce
   * allocations.
   * @param e Instance to return to object pool.
   */
  static void release(PropertyChangedEvent e) {
    _objectPool.add(e);
  }
}

File: plugin/com.hello.uxml.tools.ui/src/com/hello/uxml/tools/framework/PropertyChangedEvent.java
package com.hello.uxml.tools.framework;

import com.hello.uxml.tools.framework.events.EventArgs;

/**
 * Encapsulates property change details for listeners.
 *
 * @author ferhat
 */
public class PropertyChangedEvent extends EventArgs {

  @SuppressWarnings("unused")
  private static final long serialVersionUID = 0;

  private PropertyDefinition propDef;
  private Object oldValue;
  private Object newValue;

  /**
   * Constructs <code>PropertyChangeEvent</code> object
   * @param source Source of property change event
   * @param propDef property definition
   * @param oldValue Old property value
   * @param newValue New property value
   */
  public PropertyChangedEvent(Object source, PropertyDefinition propDef, Object oldValue,
      Object newValue) {
    super(source, null);
    this.propDef = propDef;
    this.oldValue = oldValue;
    this.newValue = newValue;
  }

  /**
   * Returns property that was changed.
   */
  public PropertyDefinition getProperty() {
    return propDef;
  }

  /**
   * Returns value of property before the change.
   */
  public Object getOldValue() {
    return oldValue;
  }

  /**
   * Returns value of property after the change.
   */
  public Object getNewValue() {
    return newValue;
  }
}
